flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java

```diff
@@ -360,11 +362,27 @@ List<String> registerProvidedLocalResources() {
                     envShipResourceList.add(descriptor);
 
                     if (!isFlinkDistJar(filePath.getName()) && !isPlugin(filePath)) {
-                        classPaths.add(fileName);
+                        URI parentDirectoryUri = new Path(fileName).getParent().toUri();
+                        String relativeParentDirectory =
```

Comment:
```
-Dyarn.provided.lib.dirs="hdfs://myhdfs/tmp/flink-1.18-SNAPSHOT/lib;hdfs://myhdfs/tmp/flink-1.18-SNAPSHOT/plugins;hdfs://myhdfs/tmp/flink-1.18-SNAPSHOT/bin"
```

It seems that we already do the relativize in the `getAllFilesInProvidedLibDirs`. So the input `providedLibDirs` is already relative path(e.g. `bin/config.sh`, `bin/find-flink-home.sh`).

In addition, adding a non-jar file into the classpath does not take too many benefits. I suggest only add the parent directory to the classpath. The Yarn ship files have the same behavior.

Then we could have a simpler code here.
```
if (fileName.endsWith("jar")) {
    resourcesJar.add(fileName);
} else {
    resourcesDir.add(new Path(fileName).getParent().toString());
}
```

Reply:
+1 to @wangyang0918 suggestion.

Reply:
Makes sense. This way we can limit the values in the classPath also. Updated the code.